src/map/lua/luautils.cpp

```diff
@@ -5825,4 +5826,65 @@ namespace luautils
 
         return table;
     }
+
+    // Determines if a table is valid to use ipairs()
+    // Also implies #tableName reliably returns the size of the table
+    bool luautils::IsTableIpairsCompatible(const sol::optional<sol::table>& maybeSequentialTable)
+    {
+        if (!maybeSequentialTable)
+        {
+            return false;
+        }
+
+        sol::table t = *maybeSequentialTable;
+
+        // Find first natural number that isn't a key of the table
+        lua_Integer index = 1;
+        for (;;)
+        {
+            if (t.raw_get<sol::object>(index).is<sol::nil_t>())
+            {
+                // Stop at first nil (ipairs behavior). I.e. index will be 1 more than the table size for i-pairs compatible tables
+                break;
+            }
+            ++index;
+        }
+
+        // "index" is the first nil value, don't consider an empty table valid
+        if (index == 1)
+        {
+            return false;
+        }
+
+        // Ensure detected "size" has no holes in the table
+        for (lua_Integer i = 1; i < index; ++i)
+        {
+            if (t.raw_get<sol::object>(i).is<sol::nil_t>())
+            {
+                // found a hole
+                return false;
+            }
+        }
+
+        // Detect any non-numeric keys
+        for (auto& kvp : t)
+        {
+            sol::object key = kvp.first;
```

Comment:
Same as my comment about `sol::table`, I'm hoping this is a fancy and special reference type, but there could be copies here if this isn't a reference type. Look into it.